### PR TITLE
Boot the app asynchronously

### PR DESCRIPTION
**Pull request by superkhau**
Connected to #170

### CODE REVIEW

**Comment by superkhau** [comment]
@bajtos I notice `npm test` has failures (most likely due to https://github.com/strongloop/loopback-boot/issues/94 as the fail log messages look familiar)

>scaffold a new app

Should I copy the code over from https://github.com/strongloop/loopback-workspace/blob/master/test/support.js#L79-L85? I don't see any code in boot that scaffolds an app.

**Comment by superkhau** [comment]
@bajtos Review please.

File: templates/components/api-server/template/server/server.js [hunk 1]
```diff
@@ -3,19 +3,21 @@ var boot = require('loopback-boot');
...
 var app = module.exports = loopback();
 
-// Bootstrap the application, configure models, datasources and middleware.
-// Sub-apps like REST API are mounted via boot scripts.
-boot(app, __dirname);
-
 app.start = function() {
   // start the web server
   return app.listen(function() {
-    app.emit('started');
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: We must preserve the event `started`, it's used by [server/boot/explorer.js](https://github.com/strongloop/loopback-workspace/blob/f031f73f00b903204a674fcfdb6570fa0bad88cc/templates/components/api-server/template/server/boot/explorer.js#L20-L26) to print a message with Explorer URL, but only after the app was started.

File: templates/components/api-server/template/server/server.js [hunk 2]
```diff
@@ -3,19 +3,21 @@ var boot = require('loopback-boot');
...
-// start the server if `$ node server.js`
-if (require.main === module) {
-  app.start();
-}
+// Bootstrap the application, configure models, datasources and middleware.
+// Sub-apps like REST API are mounted via boot scripts.
+boot(app, __dirname, function(err) {
+  if (err) throw err;
+
+  app.emit('ready');
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: An earlier comment says "This should be implemented by loopback-boot, please remove". Boot already emits `booted` event, this is what consumer should use.

File: test/end-to-end.js [hunk 3]
```diff
@@ -140,6 +140,24 @@ describe('end-to-end', function() {
...
     });
+
+    it('emits the `ready` event when booting is complete', function() {
+      var src = FIXTURES + '/emit.js';
+      var dest = SANDBOX + '/server/boot/emit.js'
+      fs.copySync(src, dest);
+      delete require.cache[require.resolve(SANDBOX)];
+      var app = require(SANDBOX + '/server/server');
+      expect(app.readyEventEmitted).to.be.ok();
+    });
```
Comment (bajtos) [inline test/end-to-end.js]: This test is not necessary, the event `booted` is triggered by `loopback-boot`. 

However, if you feel it's valuable to write an integration test capturing the contract that a server scaffolded by the template will fire `booted` event, then I am ok with that too. Just rename `ready` to `booted` everywhere.

File: test/fixtures/async.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+module.exports = function(app, cb) {
+  // ensure boot script executes asynchronously
+  app.asyncBoot = typeof cb === 'function';
```
Comment (bajtos) [inline test/fixtures/async.js]: This does not really verify that the script are executed asynchronously. You should check that the boot process waits until the async scripts finished.

I think you should rework the two new tests and merge them into one.

```js
// async.js
module.exports = function(app, cb) {
  app.asyncBoot = typeof cb === 'function';
  process.nextTick(cb);
}

// in the test
var app = require(...);
expect(app.booting, 'app.booting').to.be.true();
app.on('booted', function() {
  expect(app.asyncBoot, 'app.asyncBoot').to.be.true();
  done(); 
});
// the test will time out if `booted` is not emitted
```

Thoughts?

File: templates/components/api-server/template/server/server.js (hunk 1 above)
Comment (superkhau) [inline templates/components/api-server/template/server/server.js]: I see, its different from `booted`. Adding back. ;)

File: templates/components/api-server/template/server/server.js (hunk 2 above)
Comment (superkhau) [inline templates/components/api-server/template/server/server.js]: I forgot about the `booted` event, removing.

File: test/end-to-end.js (hunk 3 above)
Comment (superkhau) [inline test/end-to-end.js]: Rebanubg `ready` to `booted`.

File: test/fixtures/async.js (hunk 4 above)
Comment (superkhau) [inline test/fixtures/async.js]: I agree. Merging into one simple script instead.

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@bajtos Review please.

**Comment by bajtos** [comment]
One more thing to fix before you can merge this patch: The scaffolded version of loopback-boot (`2.4.0` ATM - [source](https://github.com/strongloop/loopback-workspace/blob/f031f73f00b903204a674fcfdb6570fa0bad88cc/templates/components/api-server/component.js#L17)) does not support async boot scripts. Please bump up the version to `^2.6.5`.

Note that the version of loopback-boot in the toplevel `package.json` must remain at 1.x, that's a completely different thing.

Other than that, the change LGTM. You can land it without further review after you upgrade loopback-boot.

**Comment by bajtos** [comment]
Also please make sure the test is passing on Jenkins before merging.

**Pull request merged**